### PR TITLE
Undesign client validation

### PR DESCRIPTION
**Pull request by rynowak**
This change removes a layer of abstraction. These validators now just do what they do now without creating a bunch of intermediate objects.

**Before:**
image

**After:**
image

### CODE REVIEW

File: src/Microsoft.AspNet.Mvc.DataAnnotations/CompareAttributeAdapter.cs [hunk 1]
```diff
@@ -11,26 +10,29 @@ namespace Microsoft.AspNet.Mvc.ModelBinding.Validation
...
+
         public CompareAttributeAdapter(CompareAttribute attribute, IStringLocalizer stringLocalizer)
             : base(new CompareAttributeWrapper(attribute), stringLocalizer)
         {
-            if (attribute == null)
-            {
-                throw new ArgumentNullException(nameof(attribute));
-            }
+            _otherProperty = "*." + Attribute.OtherProperty;
+
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.DataAnnotations/CompareAttributeAdapter.cs]: Don't design-in _too_ many blank lines #perf :trollface:

File: src/Microsoft.AspNet.Mvc.ViewFeatures/Rendering/IHtmlHelper.cs [hunk 2]
```diff
@@ -371,20 +370,6 @@ public interface IHtmlHelper
...
-        /// Returns information about about client validation rules for the specified <paramref name="metadata"/> or
-        /// <paramref name="expression"/>. Intended for use in <see cref="IHtmlHelper"/> extension methods.
-        /// </summary>
-        /// <param name="metadata">Metadata about the <see cref="object"/> of interest.</param>
-        /// <param name="expression">
-        /// Expression name, relative to the current model. Used to determine <see cref="ModelMetadata"/> when
-        /// <paramref name="metadata"/> is <c>null</c>; ignored otherwise.
-        /// </param>
-        /// <returns>An <see cref="IEnumerable{ModelClientValidationRule}"/> containing the relevant rules.</returns>
-        IEnumerable<ModelClientValidationRule> GetClientValidationRules(
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.ViewFeatures/Rendering/IHtmlHelper.cs]: This method was in MVC 5 and we have fielded MVC 6 questions related to its use.

File: src/Microsoft.AspNet.Mvc.DataAnnotations/CompareAttributeAdapter.cs [hunk 3]
```diff
@@ -11,26 +10,29 @@ namespace Microsoft.AspNet.Mvc.ModelBinding.Validation
...
+        private readonly string _otherProperty;
+
         public CompareAttributeAdapter(CompareAttribute attribute, IStringLocalizer stringLocalizer)
             : base(new CompareAttributeWrapper(attribute), stringLocalizer)
         {
-            if (attribute == null)
-            {
-                throw new ArgumentNullException(nameof(attribute));
-            }
+            _otherProperty = "*." + Attribute.OtherProperty;
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.DataAnnotations/CompareAttributeAdapter.cs]: `OtherProperty` does not always have the correct value when this constructor executes. That's why `CompareAttributeWrapper` exists.

File: src/Microsoft.AspNet.Mvc.DataAnnotations/CompareAttributeAdapter.cs [hunk 4]
```diff
@@ -11,26 +10,29 @@ namespace Microsoft.AspNet.Mvc.ModelBinding.Validation
...
                 throw new ArgumentNullException(nameof(context));
             }
 
-            var errorMessage = GetErrorMessage(context);
-            var clientRule = new ModelClientValidationEqualToRule(errorMessage, "*." + Attribute.OtherProperty);
-            return new[] { clientRule };
+            context.Attributes.Add("data-val-equalto", GetErrorMessage(context));
+            context.Attributes.Add("data-value-equalto-other", _otherProperty);
+
+            if (!context.Attributes.ContainsKey("data-val"))
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.DataAnnotations/CompareAttributeAdapter.cs]: _Every_ addition to the `Attributes` must be conditional. As-is this code may override values coming from `htmlAttributes` the user passed.

File: src/Microsoft.AspNet.Mvc.DataAnnotations/ModelClientValidationEqualToRule.cs [hunk 5]
```diff
@@ -1,29 +0,0 @@
...
-// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
-
-using System;
-
-namespace Microsoft.AspNet.Mvc.ModelBinding.Validation
-{
-    /// <summary>
-    /// Represents client side validation rule that determines if two values are equal.
-    /// </summary>
-    public class ModelClientValidationEqualToRule : ModelClientValidationRule
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.DataAnnotations/ModelClientValidationEqualToRule.cs]: Why does the base `ModelClientValidationRule` class still exist?

File: src/Microsoft.AspNet.Mvc.DataAnnotations/StringLengthAttributeAdapter.cs [hunk 6]
```diff
@@ -2,33 +2,40 @@
...
             }
 
-            var errorMessage = GetErrorMessage(context);
-            var rule = new ModelClientValidationStringLengthRule(errorMessage,
-                                                                 Attribute.MinimumLength,
-                                                                 Attribute.MaximumLength);
-            return new[] { rule };
+            context.Attributes.Add("data-val-length", GetErrorMessage(context));
+            context.Attributes.Add("data-val-length-max", _max);
+            context.Attributes.Add("data-val-length-min", _min);
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.DataAnnotations/StringLengthAttributeAdapter.cs]: Need to special case `_min="0"` and `_max=int.MaxValue.ToString(CultureInfo.InvariantCulture)` as was done in `ModelClientValidationStringLengthRule`.

File: src/Microsoft.AspNet.Mvc.ViewFeatures/ViewFeatures/DefaultHtmlGenerator.cs [hunk 7]
```diff
@@ -1272,28 +1248,48 @@ internal static object GetModelStateValue(ViewContext viewContext, string key, T
...
+                return;
             }
 
             formContext.RenderedField(fullName, true);
 
-            var clientRules = GetClientValidationRules(viewContext, modelExplorer, expression);
+            modelExplorer = modelExplorer ??
+                ExpressionMetadataProvider.FromStringExpression(expression, viewContext.ViewData, _metadataProvider);
 
-            return UnobtrusiveValidationAttributesGenerator.GetValidationAttributes(clientRules);
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.ViewFeatures/ViewFeatures/DefaultHtmlGenerator.cs]: Can `UnobtrusiveValidationAttributesGenerator` be removed?

File: src/Microsoft.AspNet.Mvc.ViewFeatures/ViewFeatures/IHtmlGenerator.cs [hunk 8]
```diff
@@ -341,8 +341,9 @@ public interface IHtmlGenerator
         /// Not used directly in <see cref="HtmlHelper"/>. Exposed publicly for use in other <see cref="IHtmlHelper"/>
         /// implementations.
         /// </remarks>
-        IEnumerable<ModelClientValidationRule> GetClientValidationRules(
+        void AddValidationAttributes(
             ViewContext viewContext,
+            TagBuilder tagBuilder,
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.ViewFeatures/ViewFeatures/IHtmlGenerator.cs]: When this becomes realz, document the new parameter.

File: src/Microsoft.AspNet.Mvc.DataAnnotations/CompareAttributeAdapter.cs (hunk 4 above)
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.DataAnnotations/CompareAttributeAdapter.cs]: Suggest adding a `MergeAttribute()` method to `AttributeAdapterBase`.

**Comment by dougbu** [comment]
:watch:

**Comment by dougbu** [comment]
@rynowak definitely a worthwhile perf improvement as well as a cleaner API (as long as we can indeed to w/o `IHtmlHelper.GetClientValidationRules()`) but what issue is this PR related to?

File: src/Microsoft.AspNet.Mvc.ViewFeatures/Rendering/IHtmlHelper.cs (hunk 2 above)
Comment (Eilon) [inline src/Microsoft.AspNet.Mvc.ViewFeatures/Rendering/IHtmlHelper.cs]: Chatted with @rynowak in person, I think this is fine. We can revisit in the future if we see scenarios that we simply don't handle.

**Comment by rynowak** [comment]
@dougbu - updated and removed design tag.

File: src/Microsoft.AspNetCore.Mvc.DataAnnotations/Internal/StringLengthAttributeAdapter.cs [hunk 9]
```diff
@@ -11,24 +11,32 @@ namespace Microsoft.AspNetCore.Mvc.DataAnnotations.Internal
...
             }
 
-            var errorMessage = GetErrorMessage(context);
-            var rule = new ModelClientValidationStringLengthRule(errorMessage,
-                                                                 Attribute.MinimumLength,
-                                                                 Attribute.MaximumLength);
-            return new[] { rule };
+            MergeAttribute(context.Attributes, "data-val", "true");
+            MergeAttribute(context.Attributes, "data-val-length", GetErrorMessage(context));
+            MergeAttribute(context.Attributes, "data-val-length-max", _max);
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.DataAnnotations/Internal/StringLengthAttributeAdapter.cs]: Why are you changing the semantics here? Previous code checked equivalent of `Attribute.MaximumLength != int.MaxValue` before adding `data-val-length-max`.

File: src/Microsoft.AspNetCore.Mvc.DataAnnotations/Internal/ValidationAttributeAdapterOfTAttribute.cs [hunk 10]
```diff
@@ -32,14 +32,30 @@ public ValidationAttributeAdapter(TAttribute attribute, IStringLocalizer stringL
...
+        /// <summary>
+        /// Adds the given <paramref name="key"/> and <paramref name="value"/> into
+        /// <paramref name="attributes"/> if <paramref name="attributes"/> does not contain a value for
+        /// <paramref name="key"/>.
+        /// </summary>
+        /// <param name="attributes">The HTML attributes dictionary.</param>
+        /// <param name="key">The attribute key.</param>
+        /// <param name="value">The attribute value.</param>
+        /// <returns><c>true</c> if an attribute was added, otherwise <c>false</c>.</returns>
+        protected bool MergeAttribute(IDictionary<string, string> attributes, string key, string value)
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.DataAnnotations/Internal/ValidationAttributeAdapterOfTAttribute.cs]: Not a big deal but suggest `static` for this and its few `private` copies.

File: src/Microsoft.AspNetCore.Mvc.DataAnnotations/Internal/RangeAttributeAdapter.cs [hunk 11]
```diff
@@ -11,26 +11,32 @@ namespace Microsoft.AspNetCore.Mvc.DataAnnotations.Internal
...
+        private readonly string _max;
+        private readonly string _min;
+
         public RangeAttributeAdapter(RangeAttribute attribute, IStringLocalizer stringLocalizer)
             : base(attribute, stringLocalizer)
         {
+            // This will trigger the conversion of Attribute.Minimum and Attribute.Maximum.
+            // This is needed, because the attribute is stateful and will convert from a string like
+            // "100m" to the decimal value 100.
+            attribute.IsValid(3);
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.DataAnnotations/Internal/RangeAttributeAdapter.cs]: Is there a `const` somewhere you can use instead of the magic number?

File: src/Microsoft.AspNetCore.Mvc.DataAnnotations/Internal/StringLengthAttributeAdapter.cs (hunk 9 above)
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.DataAnnotations/Internal/StringLengthAttributeAdapter.cs]: Just something I missed. I'll add a test for this.

File: src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewFeatures/DefaultHtmlGenerator.cs [hunk 12]
```diff
@@ -1277,28 +1252,48 @@ internal static object GetModelStateValue(ViewContext viewContext, string key, T
 
         // Only render attributes if client-side validation is enabled, and then only if we've
         // never rendered validation for a field with this name in this form.
-        protected virtual IDictionary<string, object> GetValidationAttributes(
+        public virtual void AddValidationAttributes(
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewFeatures/DefaultHtmlGenerator.cs]: Wondering about exposing this in `IHtmlGenerator`. Thoughts?

Going the other way, should it be `protected virtual` like most of the helpers in this class?

In any case, should have doc comments.

File: src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewFeatures/UnobtrusiveValidationAttributesGenerator.cs [hunk 13]
```diff
@@ -1,96 +0,0 @@
...
-// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
-
-using System;
-using System.Collections.Generic;
-using System.Linq;
-using Microsoft.AspNetCore.Mvc.ModelBinding.Validation;
-
-namespace Microsoft.AspNetCore.Mvc.ViewFeatures
-{
-    public static class UnobtrusiveValidationAttributesGenerator
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewFeatures/UnobtrusiveValidationAttributesGenerator.cs]: :sob:

File: test/Microsoft.AspNetCore.Mvc.DataAnnotations.Test/Internal/CompareAttributeAdapterTest.cs [hunk 14]
```diff
@@ -62,15 +74,25 @@ public void ClientRulesWithCompareAttribute_ErrorMessageUsesDisplayName()
...
+                metadataProvider,
+                new AttributeDictionary());
 
             // Act
-            var rules = adapter.GetClientValidationRules(context);
+            adapter.AddValidation(context);
 
             // Assert
-            var rule = Assert.Single(rules);
-            // Mono issue - https://github.com/aspnet/External/issues/19
```
Comment (dougbu) [inline test/Microsoft.AspNetCore.Mvc.DataAnnotations.Test/Internal/CompareAttributeAdapterTest.cs]: Please copy this above the line that sets `expectedMessage`. That will make it much easier to clean up when / if we move to a version of Mono containing a fix. Same for other places we're currently losing this information.

File: test/Microsoft.AspNetCore.Mvc.DataAnnotations.Test/Internal/StringLengthAttributeAdapterTest.cs [hunk 15]
```diff
@@ -15,7 +16,7 @@ public class StringLengthAttributeAdapterTest
     {
         [Fact]
         [ReplaceCulture]
-        public void GetClientValidationRules_WithMaxLength_ReturnsValidationParameters_Localize()
+        public void AddValidation_WithMaxLength_AddsAttributes_Localize()
```
Comment (dougbu) [inline test/Microsoft.AspNetCore.Mvc.DataAnnotations.Test/Internal/StringLengthAttributeAdapterTest.cs]: Should test the special cases in the `StringLengthAttributeAdapter` (`0` and `int.MaxValue`).

File: test/Microsoft.AspNetCore.Mvc.DataAnnotations.Test/Internal/CompareAttributeAdapterTest.cs [hunk 16]
```diff
@@ -25,19 +26,30 @@ public void ClientRulesWithCompareAttribute_ErrorMessageUsesDisplayName_WithoutL
...
+            var expectedMessage = PlatformNormalizer.NormalizeContent(
+                    "'MyPropertyDisplayName' and 'OtherPropertyDisplayName' do not match.");
+
             var actionContext = new ActionContext();
-            var context = new ClientModelValidationContext(actionContext, metadata, metadataProvider);
+            var context = new ClientModelValidationContext(
+                actionContext,
+                metadata,
+                metadataProvider,
+                new AttributeDictionary());
```
Comment (dougbu) [inline test/Microsoft.AspNetCore.Mvc.DataAnnotations.Test/Internal/CompareAttributeAdapterTest.cs]: Add tests covering non-empty `AttributeDictionary`s e.g. one that already contains `data-val = false` and `attribute-name = "attribute-value"`. Fine to spread this out a bit between `BlahAdapterTest` classes and to _not_ cover every single attribute the adapters add.

Comment (dougbu) [inline test/Microsoft.AspNetCore.Mvc.DataAnnotations.Test/Internal/CompareAttributeAdapterTest.cs]: (Wasn't tested before because `TagBuilder` did the heavy merging.)

**Comment by dougbu** [comment]
:watch:

File: src/Microsoft.AspNetCore.Mvc.DataAnnotations/Internal/ValidationAttributeAdapterOfTAttribute.cs (hunk 10 above)
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.DataAnnotations/Internal/ValidationAttributeAdapterOfTAttribute.cs]: sure

File: src/Microsoft.AspNetCore.Mvc.DataAnnotations/Internal/RangeAttributeAdapter.cs (hunk 11 above)
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.DataAnnotations/Internal/RangeAttributeAdapter.cs]: Discussed in person and improved this comment.

File: src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewFeatures/DefaultHtmlGenerator.cs (hunk 12 above)
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewFeatures/DefaultHtmlGenerator.cs]: I don't have any particular objections.

**Comment by rynowak** [comment]
@dougbu updated and added tests

File: test/Microsoft.AspNetCore.Mvc.DataAnnotations.Test/Internal/CompareAttributeAdapterTest.cs [hunk 17]
```diff
@@ -138,15 +184,61 @@ public void ClientRulesWithCompareAttribute_ErrorMessageUsesResourceOverride()
...
+                kvp =>
+                {
+                    Assert.Equal("data-val-equalto-other", kvp.Key);
+                    Assert.Equal(kvp.Value, "*.OtherProperty");
+                });
+        }
+
+        [Fact]
+        [ReplaceCulture]
+        public void AddValidation_DoesNotTrounceExistingAttributes()
```
Comment (dougbu) [inline test/Microsoft.AspNetCore.Mvc.DataAnnotations.Test/Internal/CompareAttributeAdapterTest.cs]: These are great additions and they go beyond my original request in that you've touched all the adapters. However I don't see anything confirming unrelated attributes are left alone. Just need one or two of the adapter tests to check `attribute-name = "attribute-value"`, not all.

File: src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewFeatures/DefaultHtmlGenerator.cs (hunk 12 above)
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewFeatures/DefaultHtmlGenerator.cs]: Let's go w/ `protected virtual` unless that breaks lots of tests. In that case, leave as you've got it.

Please add doc comments.

File: test/Microsoft.AspNetCore.Mvc.DataAnnotations.Test/Internal/CompareAttributeAdapterTest.cs (hunk 17 above)
Comment (rynowak) [inline test/Microsoft.AspNetCore.Mvc.DataAnnotations.Test/Internal/CompareAttributeAdapterTest.cs]: Sorry, I didn't understand your original request. I'm not adding tests for code that doesn't exist. There isn't anything in any of this code that clears the dictionary or removes attributes.

**Comment by rynowak** [comment]
Updated again - added docs and made that method protected.

File: test/Microsoft.AspNetCore.Mvc.DataAnnotations.Test/Internal/StringLengthAttributeAdapterTest.cs (hunk 15 above)
Comment (dougbu) [inline test/Microsoft.AspNetCore.Mvc.DataAnnotations.Test/Internal/StringLengthAttributeAdapterTest.cs]: thx

File: test/Microsoft.AspNetCore.Mvc.TagHelpers.Test/TestableHtmlGenerator.cs [hunk 18]
```diff
@@ -95,12 +95,13 @@ public override IHtmlContent GenerateAntiforgery(ViewContext viewContext)
             return tagBuilder;
         }
 
-        protected override IDictionary<string, object> GetValidationAttributes(
+        public override void AddValidationAttributes(
```
Comment (dougbu) [inline test/Microsoft.AspNetCore.Mvc.TagHelpers.Test/TestableHtmlGenerator.cs]: Does this compile after the `DefaultHtmlGenerator` change to `protected` for this method?

File: test/Microsoft.AspNetCore.Mvc.ViewFeatures.Test/RemoteAttributeTest.cs [hunk 19]
```diff
@@ -499,7 +585,11 @@ private static ClientModelValidationContext GetValidationContextWithArea(string
                 },
             };
 
-            return new ClientModelValidationContext(actionContext, _metadata, _metadataProvider);
+           return new ClientModelValidationContext(
```
Comment (dougbu) [inline test/Microsoft.AspNetCore.Mvc.ViewFeatures.Test/RemoteAttributeTest.cs]: Oh noos. Restore this all-important space!

File: test/Microsoft.AspNetCore.Mvc.DataAnnotations.Test/Internal/CompareAttributeAdapterTest.cs (hunk 14 above)
Comment (dougbu) [inline test/Microsoft.AspNetCore.Mvc.DataAnnotations.Test/Internal/CompareAttributeAdapterTest.cs]: thx

File: test/Microsoft.AspNetCore.Mvc.DataAnnotations.Test/Internal/CompareAttributeAdapterTest.cs (hunk 16 above)
Comment (dougbu) [inline test/Microsoft.AspNetCore.Mvc.DataAnnotations.Test/Internal/CompareAttributeAdapterTest.cs]: thx

File: test/Microsoft.AspNetCore.Mvc.DataAnnotations.Test/Internal/CompareAttributeAdapterTest.cs (hunk 17 above)
Comment (dougbu) [inline test/Microsoft.AspNetCore.Mvc.DataAnnotations.Test/Internal/CompareAttributeAdapterTest.cs]: :ok:

**Comment by dougbu** [comment]
:shipit: if everything builds locally. I noted one [suspicion](https://github.com/aspnet/Mvc/pull/3962#discussion-diff-50792987) on that.
